llvm.proj

```diff
@@ -0,0 +1,32 @@
+<Project>
+  <Import Project="Directory.Build.props" />
+  <Import Project="Directory.Build.targets" />
+
+  <Target Name="Build">
+    <ItemGroup>
+      <_LlvmBuildArg Include="-DCMAKE_BUILD_TYPE=Release" />
+      <_LlvmBuildArg Include="-DLLVM_INCLUDE_TESTS:BOOL=OFF" />
+      <_LlvmBuildArg Include='-DCMAKE_C_FLAGS="-I../llvm/include -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -DNDEBUG -D__NO_CTYPE_INLINE -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"' />
+      <_LlvmBuildArg Include='-DCMAKE_CXX_FLAGS="-I../llvm/include -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -DNDEBUG -D__NO_CTYPE_INLINE -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS"' />
+      <_LlvmBuildArg Include='-DLLVM_TARGETS_TO_BUILD="X86%3BAArch64%3BARM"' />
+    </ItemGroup>
+    <MakeDir Directories="$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot" Condition="!Exists('$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot')" />
+    <Exec WorkingDirectory="$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot" Command="cmake $(MSBuildProjectDirectory)/llvm/ @(_LlvmBuildArg->'%(Identity)',' ')" Condition="!Exists('$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot\CMakeCache.txt')" IgnoreStandardErrorWarningFormat="true" />
+    <Exec WorkingDirectory="$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot" Command="make all -j8" IgnoreStandardErrorWarningFormat="true" Condition="$(OS) != 'Windows_NT'" />
+    <Exec WorkingDirectory="$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot" Command="msbuild LLVM.sln /p:Configuration=Release" IgnoreStandardErrorWarningFormat="true" Condition="$(OS) == 'Windows_NT'" />
```

Comment:
This fails, on Windows. Is `msbuild` not in $PATH? Is there a better way to do this? I assume since we're running in a `dotnet build` context here, we can't just import a VS2017-based sln and be on our way, it needs to shell out to the build system. Maybe a call to vsvarsall? 🤷‍♂

Reply:
Not that familiar with how all this work so will just give some initial input. When we build Mono on Windows we setup the correct environment when building from command line using setup-vs-msbuild-env.bat and then building mono.sln that will trigger the full LLVM build using build-external-llvm.vcxproj that then uses build-external-llvm.bat to configure and build LLVM using target VS environment (2015/2017/2019). Since you are already running inside msbuild and use other exec tasks to configure, one option could be to build the LLVM.sln using MSBuild task instead of using Exec, using something like:

<MSBuild Projects="$(MSBuildProjectDirectory)\artifacts\tmp\BuildRoot\LLVM.sln" Targets="Build" Properties="Platform=x64;Configuration=Release" Condition="$(OS) == 'Windows_NT'"/>

Once you have it building it might be worth to cross check how we currently configure LLMV on Windows compared to how it is done above. Yet another alternative could be to reuse the already existing build scripts from Mono repro, or maybe the idea is not to bring this over to new repository and that's why we redo the LLVM builds?

Reply:
Is the error realy that it can't find msbuild or was that just an assumption? How do we setup to use a VS cmake generator making sure we get a LLVM.sln to begin with?

Reply:
If you look here https://github.com/mono/mono/blob/06673e723ec3d59b071a1b763680f252b71c5de4/msvc/build-external-llvm.bat#L288 you will see how we run cmake when build LLVM in Mono repro.

Reply:
Depending on environment we will use ninja (if found in env) or the VS version corresponding to setup environment.

Reply:
Looks like you already have needed build scripts for this in the repro:
https://github.com/dotnet/llvm-project/tree/0a3f49c40d9ec7bf3ac584eee94cfcde6491c98a/llvm/scripts/ci

Believe the best and simplest way for you would be to use those. I believe the corresponding build scripts in Mono repro is a little more updated with VS2019 support, so maybe it is worth upgrading them as well.

Reply:
Arcade is the Standard:tm: build tooling used within .NET Core, and if we want to be able to build Core-mode Mono against LLVM, then LLVM needs to come as a .nupkg, not as a .zip - nupkg is the only format with which .NET Core projects can express dependencies on each other.

I'm not sure exactly what is considered "Arcade" vs "not Arcade" from the perspective of a CMake project, but certain trappings are basically mandatory, including a top-level build.sh/build.cmd with standard flags and consuming as many helpers from eng/common as possible.

Calling into scripts in `llvm/scripts/ci` rather than CMake directly is _probably_ fine, but changes to the packaging behaviour aren't optional.

I've reconciled some of that in the latest changes, so on Windows only I'm just calling out to your script, but it does lose some of the functionality which was previously there (e.g. being able to call `build.cmd --pack` after `build.cmd --build` without a needless rebuild as part of the `--pack`)